### PR TITLE
UDP added, SSH and SMTP fixes

### PR DESCRIPTION
**Pull request by gmjohnson**
UDP added. Changes to protonuke.go and report.go to add UDP reporting/simulating. Works on AWS, however i got a weird disconnect error after running the simulation for a long time.

SSH - port changed to a non-standard port due to a port binding issue. Worked on AWS with non-standard port.

SMTP - added error checking to catch the io.EOF error, which was causing the connection to remain open and timout (also preventing new connections from opening). Additionally, changed the port to 587 because port 25 often filters SMTP traffic if it looks like spam (port 587 is used by spammers to bypass firewalls). Generates much more traffic and works on AWS.

### CODE REVIEW

File: src/protonuke/smtp.go [hunk 1]
```diff
@@ -77,7 +79,7 @@ var (
 	max_size  = 131072
 	myFQDN    = "protonuke.local"
 	TLSconfig *tls.Config
-	smtpPort  = ":25"
+	smtpPort  = ":587" // nonstandard port used to evade blocking by a firewall
```
Comment (mwisely) [inline src/protonuke/smtp.go]: We should probably put this back to `25`

File: src/protonuke/smtp.go [hunk 2]
```diff
@@ -355,12 +359,14 @@ func (s *SMTPClientSession) Handler() {
 			case strings.HasPrefix(cmd, "RSET"):
 				s.mail_from = ""
 				s.rcpt_to = []string{}
-				s.addResponse("250 I forgot to remember to forget")
+				s.addResponse("250 I forgot to remember to forget")
```
Comment (mwisely) [inline src/protonuke/smtp.go]: looks like these files need some `go fmt` cleanup

File: src/protonuke/smtp.go [hunk 3]
```diff
@@ -355,12 +359,14 @@ func (s *SMTPClientSession) Handler() {
 			case strings.HasPrefix(cmd, "RSET"):
 				s.mail_from = ""
 				s.rcpt_to = []string{}
-				s.addResponse("250 I forgot to remember to forget")
+				s.addResponse("250 I forgot to remember to forget")			
 			default:
 				s.addResponse("500 unrecognized command")
+				break
```
Comment (mwisely) [inline src/protonuke/smtp.go]: control will exit the switch without the break. it can be removed

File: src/protonuke/ssh.go [hunk 4]
```diff
@@ -56,9 +56,10 @@ type sshConn struct {
 	StdoutBuf bytes.Buffer
 }
 
+// Non-default port used because there was a port binding issue with port 22
 var (
 	sshConns []*sshConn
-	PORT     = ":22"
+	PORT     = ":8080"
```
Comment (mwisely) [inline src/protonuke/ssh.go]: Same as SMTP -- this should be the default port

File: src/protonuke/smtp.go [hunk 5]
```diff
@@ -419,11 +424,16 @@ func (s *SMTPClientSession) readSmtp() (input string, err error) {
...
+				input = "QUIT"
+				err = nil
+			}
 			break
 		}
 		if strings.HasSuffix(input, suffix) {
 			break
 		}
 	}
+	log.Debugln("Read from the buffer")
```
Comment (djfritz) [inline src/protonuke/smtp.go]: This is unnecessary.

File: src/protonuke/udp.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+type udpConn struct {
+	Conn		net.Conn
+	Host		string
+
+}
+
+var (
+	udpConns []*udpConn
+	U_PORT	 = ":10001"
```
Comment (djfritz) [inline src/protonuke/udp.go]: Why port 10001?

**Comment by djfritz** [comment]
The implementation looks fine, but udp is a transport, not a protocol. Should we be implementing a udp-based protocol instead?

File: src/protonuke/ssh.go [hunk 7]
```diff
@@ -56,10 +56,9 @@ type sshConn struct {
 	StdoutBuf bytes.Buffer
 }
 
-// Non-default port used because there was a port binding issue with port 22
 var (
 	sshConns []*sshConn
-	PORT     = ":8080"
+	PORT     = ":22" // may be useful to use a non-standard port if ssh'd into a machine to avoid port binding issue
```
Comment (floren) [inline src/protonuke/ssh.go]: You could make this an argument with a default of 22... although I dislike adding command line arguments, sometimes you gotta do it.

**Comment by djfritz** [comment]
Other, bigger changes are coming to protonuke. We may revisit some of this eventually.

**Pull request closed without merging**